internal/command/workdir/statestore_config_state.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package workdir
+
+import (
+	"encoding"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	version "github.com/hashicorp/go-version"
+	tfaddr "github.com/hashicorp/terraform-registry-address"
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/configs/configschema"
+	"github.com/hashicorp/terraform/internal/plans"
+	"github.com/zclconf/go-cty/cty"
+	ctyjson "github.com/zclconf/go-cty/cty/json"
+)
+
+var _ ConfigState[StateStoreConfigState] = &StateStoreConfigState{}
+
+// StateStoreConfigState describes the physical storage format for the state store
+type StateStoreConfigState struct {
+	Type      string          `json:"type"`     // State store type name
+	Provider  *Provider       `json:"provider"` // Details about the state-storage provider
+	ConfigRaw json.RawMessage `json:"config"`   // state_store block raw config, barring provider details
+	Hash      uint64          `json:"hash"`     // Hash of portion of configuration from config files
+}
+
+// Provider is used in the StateStoreConfigState struct to describe the provider that's used for pluggable
+// state storage. The data inside should mirror an entry in the dependency lock file.
+// This is NOT state of a `provider` configuration block, or an entry in `required_providers`.
+type Provider struct {
+	Version *version.Version `json:"version"` // The specific provider version used for the state store. Should be set using a getproviders.Version, etc.
+	Source  *Source          `json:"source"`  // The FQN/fully-qualified name of the provider.
+}
+
+type Source struct {
+	Provider tfaddr.Provider `json:"-"` // For un/marshaling purposes this needs to be exported, but it's omitted from JSON
+}
+
+// TODO - move this upstream
```

Comment:
Note: we want to move these into the `hashicorp/terraform-registry-address` dependency before merging this PR